app/src/main/java/com/amaze/filemanager/filesystem/operations/CopyOperation.java

```diff
@@ -0,0 +1,266 @@
+		this.move = move;
+		this.mode = mode;
+		this.isRootExplorer = isRootExplorer;
+
+		failedFOps = new ArrayList<>();
+	}
+
+
+	@Override
+	protected boolean check() {
+		return true;
+	}
+
+	@Override
+	protected void operate() throws IOException {
+		// initial start of copy, initiate the watcher
+		watcherUtil.watch(copyService);
+
+		if (FileUtil.checkFolder((targetPath), context) == 1) {
+			for (int i = 0; i < sourceFiles.size(); i++) {
+				sourceProgress = i;
+				HybridFileParcelable f1 = (sourceFiles.get(i));
+
+				try {
+					HybridFile hFile;
+					if (targetPath.contains(context.getExternalCacheDir().getPath())) {
+						// the target open mode is not the one we're currently in!
+						// we're processing the file for cache
+						hFile = new HybridFile(OpenMode.FILE, targetPath, sourceFiles.get(i).getName(),
+								f1.isDirectory());
+					} else {
+
+						// the target open mode is where we're currently at
+						hFile = new HybridFile(mode, targetPath, sourceFiles.get(i).getName(),
+								f1.isDirectory());
+					}
+
+					if (progressHandler.getCancelled()) {
+						throw new CancellationIOException();
+					}
+
+					if ((f1.getMode() == OpenMode.ROOT || mode == OpenMode.ROOT)
+							&& isRootExplorer) {
+						// either source or target are in root
+						Log.d(getClass().getSimpleName(), "either source or target are in root");
+						progressHandler.setSourceFilesProcessed(++sourceProgress);
+						copyRoot(f1, hFile, move);
+						continue;
+					}
+					progressHandler.setSourceFilesProcessed(++sourceProgress);
+					copyFiles((f1), hFile, progressHandler);
+				} catch (IllegalStateException | IOException e) {
+					for (int j = i; j < sourceFiles.size(); j++)
+						failedFOps.add(sourceFiles.get(j));
+
+					throw e;
+				}
+			}
+
+		} else if (isRootExplorer) {
+			for (int i = 0; i < sourceFiles.size(); i++) {
+				if (!progressHandler.getCancelled()) {
+					HybridFile hFile = new HybridFile(mode, targetPath, sourceFiles.get(i).getName(),
+							sourceFiles.get(i).isDirectory());
+					progressHandler.setSourceFilesProcessed(++sourceProgress);
+					progressHandler.setFileName(sourceFiles.get(i).getName());
+					copyRoot(sourceFiles.get(i), hFile, move);
+
+                    /*
+                    if(checkFiles(new HybridFile(sourceFiles.get(i).getMode(), path),
+                            new HybridFile(OpenMode.ROOT,targetPath+"/"+name))) {
+                        failedFOps.add(sourceFiles.get(i));
+                    }
+                    */
+				}
+			}
+		} else {
+			failedFOps.addAll(sourceFiles);
+			throw new IOException("Copy is impossible!");
+		}
+
+		// making sure to delete files after copy operation is done
+		// and not if the copy was cancelled
+		if (move && !progressHandler.getCancelled()) {
+			for (HybridFileParcelable a : sourceFiles) {
+				if (!failedFOps.contains(a)) {
+					DeleteOperation delete = new DeleteOperation(context, isRootExplorer, a);
+					requires(delete);
```

Comment:
Isn't there an `undo` method that does this?

Reply:
Undo reverses the entire operation (deleting the copies), this deletes the source files.

Reply:
Ah, I see what you're getting at. This class also doubles as `MoveOperation`, correct?

Reply:
> Ah, I see what you're getting at. This class also doubles as `MoveOperation`, correct?

Yes.